text/0066-separate-context-propagation.md

```diff
@@ -0,0 +1,578 @@
+set of functions which read, write, and propagate data.
+
+**`GetBaggage(context, key) -> value`**  
+To access the distributed state of a concern, the Baggage API provides a 
+function which takes a context and a key as input, and returns a value.
+
+**`SetBaggage(context, key, value) -> context`**  
+To record the distributed state of a concern, the Baggage API provides a 
+function which takes a context, a key, and a value as input, and returns an 
+updated context which contains the new value.
+
+**`RemoveBaggage(context, key) -> context`**  
+To delete distributed state from a concern, the Baggage API provides a function 
+which takes a context, a key, and a value as input, and returns an updated 
+context which contains the new value.
+
+**`ClearBaggage(context) -> context`**  
+To avoid sending baggage to an untrusted downstream process, the Baggage API 
+provides a function to remove all baggage from a context.
+
+**`GetBaggagePropagator() -> (HTTP_Extractor, HTTP_Injector)`**  
+To deserialize the state of the system sent from the the prior upstream process, 
+and to serialize the the current state of the system and send it to the next 
+downstream process, the Baggage API provides a function which returns a 
+baggage-specific implementation of the HTTPExtract and HTTPInject functions.
+
+# Context Propagation
+
+## Context API
+
+Cross-cutting concerns access data in-process using the same, shared context 
+object. Each concern uses its own namespaced set of keys in the context, 
+containing all of the data for that cross-cutting concern.
+
+**`GetValue(context, key) -> value`**  
+To access the local state of an concern, the Context API provides a function 
+which takes a context and a key as input, and returns a value.
+
+**`SetValue(context, key, value) -> context`**  
+To record the local state of a cross-cutting concern, the Context API provides a 
+function which takes a context, a key, and a value as input, and returns an 
+updated context which contains the new value.
+
+
+### Optional: Automated Context Management
+When possible, the OpenTelemetry context should automatically be associated 
+with the program execution context. Note that some languages do not provide any 
+facility for setting and getting a current context. In these cases, the user is 
+responsible for managing the current context.  
+
+**`GetCurrent() -> context`**  
+To access the context associated with program execution, the Context API 
+provides a function which takes no arguments and returns a Context.
+
+**`SetCurrent(context)`**  
+To associate a context with program execution, the Context API provides a 
+function which takes a Context.
+
+
+
+## Propagation API
+
+Cross-cutting concerns send their state to downstream processes via propagators: 
+functions which read and write context into RPC requests. Each concern creates a 
+set of propagators for every type of supported medium - currently only HTTP 
```

Comment:
Do we see adding more formats in the near future, and if so, does it make sense to prefix these methods for each format, or do you see having a different Propagation API for each format. 

I think some psuedo code will help clarify my question. If we add another format, what would it look like:

**Option 1** - Prefixed methods

```
//HTTP Format
ctx = Propagation.HttpExtract(...)
headers = Propagation.HttpInject(...)

//Binary Format
ctx = Propagation.BinaryExtract(...)
headers = Propagation.BinaryInject(...)
```

**Option 2** Propagation Object per Format

```

//HTTP Format
ctx = HttpPropagation.Extract(...)
headers = HttpPropagation.Inject(...)

//Binary Format
ctx = BinaryPropagation.Extract(...)
headers = BinaryPropagation.Inject(...)
```

or something else?

Reply:
I suppose there is an **Option 3** where we manage different sets of global injectors / extractors, but keep the same inject and extract methods. This works so long as headers is a reasonable return type for injecting the new format.

```
//HTTP
ctx = Propagation.Extract(context, Propagation.GetHttpExtractors, headers)
headers = Propagation.Inject(context, Propagation.GetHttpInjectors, headers)

//Binary
ctx = Propagation.Extract(context, Propagation.GetBinaryExtractors, headers)
headers = Propagation.Inject(context, Propagation.GetBinaryInjectors, headers)
```

Reply:
I personally like Option 1, as you can have the propagation options passed to the user as a single object, and he could *then* decide what format to go with (I'm also thinking we won't be adding many formats in the near future, other than **http** and **binary**).

Reply:
@mwear I definitely think they need to be namespaced. I'm skeptical of writing it as **Option 3**, since that involves polymorphism. It would be better to keep each format it's own thing.

I personally like `Option 2`, where it is possible to add new formats without perturbing the existing APIs.

Reply:
(but I also think it doesn't matter too much! All three options would be acceptable, assuming the language let you write it that way; the necessary bits are all there.)

Reply:
I like option #1. See my related remark here: https://github.com/open-telemetry/oteps/pull/66#discussion_r350561807